support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -233,3 +251,109 @@ object CreateZuoraSubscription {
     }
 
 }
+
+object DigitalSubscriptionGiftRedemption {
+
+  sealed trait SubscriptionState
+
+  case class Unredeemed(subscriptionId: String) extends SubscriptionState
+
+  // This can happen if Zuora is responding very slowly - a redemption request may succeed but not return a response
+  // until after the CreateZuoraSubscription lambda has timed out meaning that the redemption will be retried with the
+  // same requestId. In this case we want the lambda to succeed so that we progress to the next lambda
+  case object RedeemedInThisRequest extends SubscriptionState
+
+  case object Redeemed extends SubscriptionState
+
+  case object NotFound extends SubscriptionState
+
+  def maybeDigitalSubscriptionGiftRedemption(product: ProductType, paymentMethod: Either[PaymentMethod, RedemptionData]) = {
+    product match {
+      case d: DigitalPack if paymentMethod.isRight && d.readerType == Gift => paymentMethod.right.toOption
+      case _ => None
+    }
+  }
+
+  def redeemGift(
+    redemptionData: RedemptionData,
+    requestInfo: RequestInfo,
+    state: CreateZuoraSubscriptionState,
+    zuoraService: ZuoraService,
+    catalogService: CatalogService
+  ): Future[HandlerResult[SendThankYouEmailState]] =
+    zuoraService.getSubscriptionFromRedemptionCode(redemptionData.redemptionCode).flatMap(
+      redemptionQueryResponse =>
+        getSubscriptionState(redemptionQueryResponse, state.requestId.toString) match {
+          case u: Unredeemed => doRedemption(u.subscriptionId, state, redemptionData, requestInfo, zuoraService, catalogService)
+          case RedeemedInThisRequest => Future.fromTry(buildHandlerResult(UpdateRedemptionDataResponse(true), state, redemptionData, requestInfo))
+          case Redeemed => Future.fromTry(Failure(new RuntimeException(GetCodeStatus.CodeAlreadyUsed.clientCode)))
+          case NotFound => Future.fromTry(Failure(new RuntimeException(GetCodeStatus.NoSuchCode.clientCode)))
+        }
+    )
+
+  private def getSubscriptionState(existingSub: SubscriptionRedemptionQueryResponse, requestId: String) = {
+    existingSub.records.headOption.map(
+      existingSubFields =>
+        if (existingSubFields.gifteeIdentityId.isEmpty)
+          Unredeemed(existingSubFields.id)
+        else if (existingSubFields.createdRequestId == requestId)
+          RedeemedInThisRequest
+        else Redeemed
+    ).getOrElse(NotFound)
```

Comment:
I'd slightly prefer a pattern match here, for a runtime assertion of one result but it also helps with clarity a bit 
```suggestion
    existingSub.records match {
      case existingSubFields :: Nil if (existingSubFields.gifteeIdentityId.isEmpty) =>
        Unredeemed(existingSubFields.id)
      case existingSubFields :: Nil if (existingSubFields.createdRequestId == requestId) =>
        RedeemedInThisRequest
      case existingSubFields :: Nil =>
        Redeemed
      case _ =>
        NotFound
    }
```